---
title: "Comprehensive Testing of pgp_rcpp()"
output: html_notebook
---

Here I just create some data sets that are comprehensive, in terms of the number
of combinations of different variables that are being fed into the thing, so I can
inspect the results at the different debug levels of pgp_rcpp().

## A set of test variables

I will do three chromosomes, each with enough intersected segments to span
the whole range of possibilities (with three ancestries) on each chromosome.
Each segment will have diagnostic markers and then variable markers that cover the
range of possible genotypes in the pair of individuals.  I will have a candidate kid and
two candidate parents.

So, it will look something like this:
```{r}
library(tidyverse)
IXtib <- expand_grid(
  kIdx = 0L,
  pIdx = c(1L, 2L),
  cIdx = c(0L, 1L),
  anck = as.integer(c(2, 4, 6, 10, 12, 18)),
  ancp = as.integer(c(2, 4, 6, 10, 12, 18))
) %>%
  mutate(
    start = seq(1e5L, by = 1e5L, length.out = n()),
    stop = seq(2e5L, by = 1e5L, length.out = n()) - 1L
  ) %>%
  expand_grid(
    isDiag = 0:1,
    gk = -1:2,
    gp = -1:2
  ) %>%
  group_by(pIdx) %>%
  mutate(
    mIdx = (1:n()) - 1L, 
    .before = gk
  ) %>%
  group_by(start, stop) %>%
  mutate(
    pos = start + seq(100L, by = 100L, length.out = n()),
    .after = mIdx
  ) %>%
  ungroup()
  
```

After that, we need the admixture fractions for everyone.  There are 3 individuals (0, 1, and 2)
currently, so we make that many.
```{r}
ADmat <- matrix(
  c(
    0.2, 0.7, 0.1,
    0.4, 0.6, 0,
    0.7, 0, 0.3
  ),
  byrow = TRUE,
  ncol = 3
)
```


Here is a simple function to make a matrix of allele freqs, one for each
marker in the IXtib.  Every marker is going to get the same thing here.  
```{r}
#' @param freqs alle freqs for the different ancestries.  length of this is the
#' number of ancestries.
makeFreqs <- function(IXtib, freqs = c(0.5, 0.5, 0.5)) {
  m <- max(IXtib$mIdx) + 1L
  
  matrix(freqs, byrow = TRUE, ncol = length(freqs), nrow = m)
}


```

So, the we can run this like the following:

## Debug == 1, check the subscripting of the markers within each segment and the haplotypes

```{r}
IXG <- IXtib %>% select(-isDiag) %>% as.matrix()
AF <- makeFreqs(IXtib)
isD <- IXtib %>% filter(kIdx == 0, pIdx == 1) %>% pull(isDiag)
AD <- ADmat
debug <- 1L  

pgp_result <- pgp_rcpp(IXG, AF, isD, AD, as.integer(debug)) %>% 
  as_tibble()

pgp_result
```

If you then do `View(pgp_result)` and compare it to the following, you see that
it checks out:
```{r}
IXtib %>%
  group_by(kIdx, pIdx, cIdx, start, stop) %>% 
  slice(1, n())
```
You can see that the indexes are lining up correctly with lo and hi.

Also, inspecting pgp_result shows that the the vectors made from the trits are correct:
```{r}
pgp_result %>%
  count(anc_k, anc_p, haplist_k, haplist_p)
```

So, everything is checking out on debug == 1.  Let's continue sussing things out.


## Debug ==  2.  Checking up on the genotype probabilities given the kid is unrelated to the parent

For this, we are going to whittle it down to just a single chromosome and a single parent, which
will still have all the different ancestries and the genotypes.
```{r}
IXshort <- IXtib %>%
  filter(pIdx == 1, cIdx == 0) %>%
  mutate(mIdx = 0:(n()-1L))


IXG <- IXshort %>% select(-isDiag) %>% as.matrix()
isD <- IXtib %>% filter(kIdx == 0, pIdx == 1) %>% pull(isDiag)
AD <- ADmat
debug <- 2L  
```


### First sanity check: same allele freqs of 0.5 in all ancestries


```{r}
AF <- makeFreqs(IXshort)

sc1 <- pgp_rcpp(IXG, AF, isD, AD, as.integer(debug)) %>% 
  as_tibble()

# See that the genotype probs are correct
sc1 %>%
  count(par_geno, kid_geno, geno_prob)

```

Now, let's check the SOS probs.  Recall the admixture fractions look like this:
```{r}
ADmat[1,]
```

And we can see what the SOS probs look like for the kid:
```{r}
sc1 %>%
  count(anc_k, prob_k_SOS) %>%
  mutate(cumul = cumsum(prob_k_SOS))
```
That all checks out there.  Cool.

### Sanity check 2.  Genotype frequencies spot check

We set the genotype freqs to different values in those ancestries and
check them.
```{r}
AF <- makeFreqs(IXshort, freqs = c(0.2, 0.3, 0.4))

sc2 <- pgp_rcpp(IXG, AF, isD, AD, as.integer(debug)) %>% 
  as_tibble()

# See that the genotype probs are correct
sc2_1 <- sc2 %>%
  count(anc_k, par_geno, kid_geno, geno_prob)
```

One simple check is that the genotype probabilities sum to one within each ancestry:
```{r}
sc2_1 %>%
  filter(kid_geno != -1 & par_geno != -1) %>% 
  group_by(anc_k, par_geno) %>%
  summarise(tot_prob = sum(geno_prob))
```

That is encouraging.  Recall the freqs are:
```{r}
AF[1,]
```
Now we can spot check things:
```{r}
sc2_1 %>%
  filter(kid_geno != -1 & par_geno != -1) %>%
  count(anc_k, kid_geno, geno_prob)
  
```

```{r}
# anc = 4 geno = 1 should be:
0.8 * 0.3 + 0.2 * 0.7

# anc = 4 geno = 2 should be:
0.2 * 0.3

# anc = 10 geno = 1 should be:
0.8 * 0.4 + 0.2 * 0.6

# anc = 10 geno = 2 should be:
0.2 * 0.4
```
Yes!  This stuff all checks out. Cool.  It is now time to proceed to the probabilities
under the parent-offspring hypothesis, which necessarily involves cases when
different ancestries are being segregated from the parent.  


## Debug == 3. Probabilities conditional on the segregated haplotypes from both the parent and the population

Here we will check out the SOS probabilities.  

First, run it with debug == 3
```{r}
debug <- 3L
sc_d3_1 <- pgp_rcpp(IXG, AF, isD, AD, as.integer(debug)) %>% 
  as_tibble()
```


### A quick sanity check test

I am going to put one variable marker in per intersected segment and I will
change the genotypes. I should be able to do that like this:
```{r}
IXtib2 <- expand_grid(
  kIdx = 0L,
  pIdx = c(1L),
  cIdx = c(0L),
  gk = -1:2,
  gp = -1:2,
  anck = as.integer(c(2, 4, 6, 10, 12, 18)),
  ancp = as.integer(c(2, 4, 6, 10, 12, 18)),
  isDiag = 0:1
) %>%
  mutate(
    start = seq(1e5L, by = 1e5L, length.out = n()),
    stop = seq(2e5L, by = 1e5L, length.out = n()) - 1L
  ) %>%
  group_by(pIdx) %>%
  mutate(
    mIdx = (1:n()) - 1L, 
    .before = gk
  ) %>%
  group_by(start, stop) %>%
  mutate(
    pos = start + seq(100L, by = 100L, length.out = n()),
    .after = mIdx
  ) %>%
  ungroup()

IXtib2 <- IXtib2[, names(IXtib)]
```

Then proceed with all allele frequencies equal to 0.5 in all populations, 
so the admixture does not factor into things at all for variable markers.
```{r}
IXG2 <- IXtib2 %>% select(-isDiag) %>% as.matrix()
AF2 <- makeFreqs(IXtib2)
isD <- IXtib2 %>% filter(kIdx == 0, pIdx == 1) %>% pull(isDiag)
AD <- AD[1:2,]
debug <- 3L  

pgp_result <- pgp_rcpp(IXG2, AF2, isD, AD, as.integer(debug)) %>% 
  as_tibble()

gps_included <- pgp_result %>%
  mutate(
    prob_diag = exp(LogProbDiag),
    prob_var = exp(LogProbVar)
  ) %>%
  left_join(., IXtib2 %>% select(start, stop, gk, gp, anck, ancp, isDiag), by = join_by(start, stop, anc_k == anck, anc_p == ancp)) 

gps_included %>%
  count(isDiag, gp, gk, prob_var) %>%
  arrange(prob_var)
```
That all checks out. 

Now, look at that for the diagnostic ones.
```{r}
gps_included %>%
  count(isDiag, gp, gk, prob_diag) %>%
  arrange(prob_diag)
```
That looks like it should.  OK.  So, maybe we are ready to try to finish up the
function.
