---
title: "A description of the MixedUpParents method"
author: "Eric C. Anderson"
output: pdf_document
date: "2023-06-22"
---

\newcommand{\pa}{{\mathrm{par}}}
\newcommand{\kid}{{\mathrm{kid}}}
\newcommand{\ppp}{\psi}
\newcommand{\opro}{\omega}


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Herein we describe a method for conducting parentage inference within populations
of individuals that are admixed between up to 3 species or subspecies.  The
particular use case we have in mind is one in which parentage inference is to be
done in populations with individuals admixed between westslope cutthroat trout,
rainbow trout, and, to a much lesser degree, if at all, yellowstone cutthroat trout.
However, while describing it
we will simply refer to three different subspecies or ancestries: $A$, $B$, and $C$.
Note that we could expand the number of different subspecies if desired, but we
are focusing on three at this point because of our particular use case.


What is interesting about this problem is that there are two sets of markers.  The
first set of markers are what we will call the "diagnostic" markers.  These are
markers that carry one allele that is only found in one of the subspecies, in which it
is fixed or nearly fixed, with the
other allele (or alleles) being found in the remaining two subspecies.  Thus, there
are diagnostic markers for $A$, and a separate set of diagnostic markers for $B$, and
a third that are diagnostic for $C$.

The second set of markers are those that segregate variation in one or more of the
subspecies.  We can call these the "variable" markers.  It is worth pointing out
that both of these sets of markers are useful for parentage inference in admixed
populations.

For doing parentage inference we do a two-step approach that makes it relatively
easy to account for chromosomal segments of distinct ancestry in the individuals
under study.

In the first step, we use the diagnostic markers to identify genomic regions that
carry $A$, $B$, or $C$ ancestry upon 0, 1, or 2 chromosomes within each sampled
individual.  In the second step, we calculate the likelihood ratio for parent-offspring vs unrelated. This
step is done conditional on the inferred number of copies of different ancestry,
at both the candidate parent and offspring, at the location of each
of the variable markers.  We will explain each of
these steps (and several other necessary steps) in the following sections.


# Inferring ancestral copy-number segments

In order to infer the copy-number (0, 1, or 2) of different ancestries ($A$, $B$, or $C$)
in genomic segments throughout the genome we use a simple heuristic rule to combine
runs of diagnostic marker genotypes into segments. Taking the $A$-diagnostic markers
for an example, at each $A$-diagnostic marker we observe either 0, 1, or 2 copies of the $A$-diagnostic
allele.  After removing missing data, any two adjacent $A$-diagnostic markers,
on the same chromosome, with the
same number of copies of $A$-diagnostic alleles define an $A$-diagnostic segment
with that number of copies.  For example, if two adjacent markers each show 2 copies
of the $A$-diagnosic allele, then the segment between them is identified as having
2 copies of $A$ ancestry.  A similar process is used to define $B$- and
$C$-diagnostic segments.

At the end of that first step we have a number of segments within an individual.
We will let $A_x$, $x\in\{0,1,2\}$ denote the set of intervals within the genome
at which the individual is identified to have $x$ copies of $A$ ancestry. $B_x$ and
$C_x$ are sets defined similarly for $B$ and $C$ ancestry.  

Because the diagnostic markers for some
subspecies might not be as widespread as those for another, it can
be advantageous to use, for example, the results for
$B$ and $C$ to help infer $A$ segments as well---in other words, if you are at part of the
genome where you don't have many $A$ markers, but you do have $B$- and $C$-diagnostic
markers, and you see that you have zero $B$-diagnostic and zero $C$-diagnostic alleles at
those markers, then
it is reasonable to assume that you have 2 copies of $A$ there.  By this method we can
extend the length of intervals of inferred ancestry copy number.
While we are doing that we can also remove segments that are contradictory
in terms of their ancestry copy numbers---for example, those that have 2 copies
of a $C$-diagnostic allele but also 1 copy of a $B$-diagnostic allele, though, in
the trout data, we found very few of these contradictory segments.

We make these extended segments and also identify contradictory segments
using interval arithmetic as implemented
in the R package 'bedtools.' Though that package is not particularly fast, the operations
can be parallelized over individuals which allows us to handle the thousands of of
trout in the data set.
Let $X \cup Y$ denote the union of sets of intervals, $X$ and $Y$, and $X \cap Y$
denote the interval-intersection of those same sets of intervals (i.e., all the overlapping
segments between the intervals in $X$ and in $Y$), and $X - Y$ denote
the set-difference (all the intervals in $X$ after removing any overlaps with $Y$).
With that notation we can easily write down rules for the sets of ancestry-copy-number
intervals after "extending" them by using available information from the other
subspecies. Let $A_xB_yC_z$ denote
the extended copy-number intervals in which the individual carries $x$ copies of $A$,
$y$ copies of $B$, and $z$ copies of $C$.  Since these are diploid organisms any
allowable value of $A_xB_yC_z$ must have $x,y,z \in \{0, 1, 2\}$ and $x + y + z = 2$.
We can find these segments with interval arithmetic as follows:
$$
\begin{aligned}
A_1B_1C_0 &= \biggl(A_1 \cup \bigl((B_1 \cap C_0) - (A_0 \cup A_2)\bigr)\biggr) \cap 
\biggl(B_1 \cup \bigl((A_1 \cap C_0) - (B_0 \cup B_2)\bigr)\biggr) \\
A_1B_0C_1 &= \biggl(A_1 \cup \bigl((B_0 \cap C_1) - (A_0 \cup A_2)\bigr)\biggr) \cap 
\biggl(C_1 \cup \bigl((A_1 \cap B_0) - (C_0 \cup C_2)\bigr)\biggr) \\
A_0B_1C_1 &= \biggl(B_1 \cup \bigl((A_0 \cap C_1) - (B_0 \cup B_2)\bigr)\biggr) \cap 
\biggl(C_1 \cup \bigl((A_0 \cap B_1) - (C_0 \cup C_2)\bigr)\biggr) \\
A_2B_0C_0 &= A_2 \cup \bigl((B_0 \cap C_0) - (A_0 \cup A_1)\bigr) \\
A_0B_2C_0 &= B_2 \cup \bigl((A_0 \cap C_0) - (B_0 \cup B_1)\bigr) \\
A_0B_0C_2 &= C_2 \cup \bigl((A_0 \cap B_0) - (C_0 \cup C_1)\bigr) \\
\end{aligned}
$$

In addition to the operations above, we do a final step of merging intervals with the same ancestry
on the same chromosome in the same individual if they are within 8 bp of one another.
This removes the holes that can occur when a single species-diagnostic marker
in a run of one particular ancestry has experienced a genotyping error.

## Implementation Notes

I have implemented the above in the function `extend_ancestral_segments_3()`. It uses the
'intervals' package.  It is not super fast, but we can parallelize it using the
'furrr' package, which makes it tolerable for operations on individuals. (But I note that
when we start doing operations on all pairs of individuals we are going to need to
write our own code with RCpp to make it fast enough...)


I have checked the output of this function and it looks pretty right on.  I have put
it on the old figure we had of the doses of different diagnostic segments.  The top
bar, surrounded by a thin black line gives the full ancestral copy number segments
after extension. Figure 1 shows that. You will have to zoom
in quite a bit, but you can do that easily since this is a PDF.

```{r fish10, echo=FALSE, fig.cap='Copy number segments and extended segments from 10 fish.'}
knitr::include_graphics("figures/E-gplot.pdf")
```


# Estimating allele frequencies

With the genomes of all the individuals in a population segmented into the
six ancestry-copy-number categories above, we can estimate the frequencies of the
_variable_-marker alleles on each of the three backgrounds, A, B, or C, within the
population, using an EM-algorithm. 

Consider a single variable-marker locus with alleles $a_i, a_j,\ldots$. Let $n_{xyz}(a_ia_j)$ denote
the number of individuals that have variable marker genotype $a_ia_j$ and also in which this locus occurs within an $A_xB_yC_z$ segment. 
In many cases, it is straightforward to count the number of alleles of different types that
must occur on a particular background (i.e., on an $A$, $B$, or $C$ background). For example, if we
are counting the number of $a_i$ alleles that occur on the $A$ background, there are many individual
genotypes in which we can be certain that the $a_i$ allele occurs on the $A$ backgound, and we can
count them directly.  In fact, the number of $a_i$ alleles on the $A$ background that we are certain
of can be expressed as:
$$
2n_{200}(a_ia_i) + n_{110}(a_ia_i) + n_{101}(a_ia_i) + \sum_{j\neq i} n_{200}(a_ia_j),
$$
The only scenarios that are
ambiguous about how many $a_i$ alleles are on
the $A$ background are ones in which the individual is heterozygous at the locus and
has only one copy of the $A$ ancestry---namely the individuals counted by
$$
n_{110}(a_ia_j)~~~\mbox{and}~~~n_{101}(a_ia_j)~~~\forall j\neq i
$$
To incorporate the observations of alleles in such individuals we use the
EM algorithm to find the maximum likelihood estimate of allele frequencies
on different ancestry backgounds.  To do so we can count up "fractional" parts
alleles according to the posterior probability that they are on the
$A$ (or $B$ or $C$) background given the current estimates of the frequency of those alleles on
each background. Let $p_A(a_i)$, $p_B(a_i)$, and $p_C(a_i)$ be the
current estimates for the frequencies of allele $a_i$ on each of the three ancestry
backgrounds. Then, for the $n_{110}(a_ia_j)$ and $n_{101}(a_ia_j)$ cases,
_a priori_ (i.e., without looking at the allelic type) the probability that the
$a_i$ allele is on the $A$-background is $\frac{1}{2}$, and with probability $\frac{1}{2}$
it is on the $B$ or $C$ background respectively.  So, the posterior probability
that the $a_i$ is on the $A$ background in each case is:
$$
\begin{aligned}
q^{(A)}_{110}(a_i;a_j) &=\frac{p_A(a_i)p_B(a_j)}{p_A(a_i)p_B(a_j) + p_A(a_j)p_B(a_i)}~~~~,~~~~\mbox{for the}~n_{110}(a_i, a_j)~\mbox{case} \\
q^{(A)}_{101}(a_i;a_j) &=\frac{p_A(a_i)p_C(a_j)}{p_A(a_i)p_C(a_j) + p_A(a_j)p_C(a_i)}~~~~,~~~~\mbox{for the}~n_{101}(a_i, a_j)~\mbox{case} \\
\end{aligned}
$$
Thus, given current estimates of $p_A(a_i)$, $p_B(a_i)$, and $p_C(a_i)$, we find new estimates
by first conducting the Expectation or "E" step---counting up the number of alleles (including "fractional" alleles), for each $a_i$:
$$
\begin{aligned}
N^\prime_A(a_i) &= 2n_{200}(a_ia_i) + n_{110}(a_ia_i) + n_{101}(a_ia_i) + \sum_{j\neq i} \biggl(n_{200}(a_ia_j) + q^{(A)}_{110}(a_i;a_j)n_{110}(a_ia_j) + q^{(A)}_{101}(a_i;a_j)n_{101}(a_ia_j)  \biggr) \\
N^\prime_B(a_i) &= 2n_{020}(a_ia_i) + n_{110}(a_ia_i) + n_{011}(a_ia_i) + \sum_{j\neq i} \biggl(n_{020}(a_ia_j) + q^{(B)}_{110}(a_i;a_j)n_{110}(a_ia_j) + q^{(B)}_{011}(a_i;a_j)n_{011}(a_ia_j)  \biggr) \\
N^\prime_C(a_i) &= 2n_{002}(a_ia_i) + n_{101}(a_ia_i) + n_{011}(a_ia_i) + \sum_{j\neq i} \biggl(n_{002}(a_ia_j) + q^{(C)}_{101}(a_i;a_j)n_{101}(a_ia_j) + q^{(C)}_{011}(a_i;a_j)n_{011}(a_ia_j)  \biggr) \\
\end{aligned}
$$
With those estimate counts in hand, we do the Maximization or "M" step, finding
the maximum likelihood estimate of the $p$'s:
$$
p^\prime_X(a_i) = \frac{N^\prime_X(a_i)}{\sum_k N^\prime_X(a_k)}~~~\mbox{for}~X\in\{A,B,C\}
$$
Iterating the E and M steps to convergence yields the maximum likelihood estimates of the
allele freqencies on the different ancestry backgrounds.


# Contributions of the species diagnostic markers to parentage inference

There are two ways that the diagnostic markers can...

# Parentage likelihood calculations with the polymorphic markers

We do the parentage likelihood calculations conditional on the estimated ancestry-copy-numbers of
the pair of individuals.  For each possible pair of a candidate parent and a candidate offspring
we first take the interval intersection of the extended ancestry segments found in each individual.
While doing so, we keep track of the $A_xB_yC_z$ of each individual at each location in the
intersection.  Some of those intersecting regions will be incompatible between the pair.  For example,
if the candidate parent has an $A_1B_1C_0$ and that candidate offspring is an $A_0B_0C_2$, then
that would be incompatible.  We will keep track of the total genomic length of incompatible intersected
segments, and also, probably, the total number of markers within those incompatible segments, and we
can use that to screen out pairs that are very unlikely to be parent-offspring. 

Ultimately, we will keep all the intersected intervals that are compatible with a parent-offspring
relationship and we will calculate the joint probability of the pair of genotypes at all the
variable markers within each
each such interval under the parent-offspring hypothesis and the unrelated hypothesis.  
Just a side note here, I think that this is going to require using RCpp if it is to
be sufficiently fast.

We will assume that each of these compatible, intersected intervals was inherited independently as
a single chunk from the parent to the kid.  This is a pretty reasonable assumption---the only place
it gets dodgy is if you have lots of missing data, and hence multiple disjoint intervals that really
should be a single, connected interval.  But otherwise, this approach does a good job of accounting for
the fact that markers are inherited in big blocks.  

We compute the joint genotype probability of the pair as a sum over the different possible
gametes that could be segregated by the parent. 
If we assume no LD of the variable markers within the different ancestries (a fairly reasonable assumption)
then the joint genotype probability is simply the product over variable markers within each intersected
interval, inside that sum over segregated ancestries.  

We step back here and sketch this out in mathematical form---the joint probability of
the candidate parent and offspring genotypes can be computed using the following sort
of scheme:
$$
\begin{aligned}
P(Y_P, Y_C) &= \prod_{{\mathrm{compatible~inter-} \atop \mathrm{sected~intervals}}}
\sum_{{ \mathrm{ancestry~segment} \atop \mathrm{segregated,~}G }}
\prod_{{\mathrm{variable~markers} \atop \mathrm{in~segment} }}
\sum_{{ \mathrm{variable~alleles~}v \atop \mathrm{assigned~to~}G,G'}}\biggl[ \\
\mbox{}& \mbox{Prob(parent genotype given }v\mbox{ on }G\mbox{ and }G') \times \\
\mbox{}& \mbox{Prob(offsping genotype given }v\mbox{ on }G\mbox{ was segregated to it}) \\
\biggr]
\end{aligned}
$$


To understand this, it is worth considering
all the different possibilities.  Let $H_1$ and $H_2$ denote two distinct ancestries/subspecies.
Note that within a single individual, in a single intersected interval,
you will never find more than two.
If an individual carries $H_1H_1$ then you know with probability 1 it will
segregate a gamete of ancestry $H_1$.  However, if it is $H_1H_2$, then you know
that with probability $\frac{1}{2}$ it segregates an $H_1$, and with probability
$H_2$ it segregates an $H_2$.  


Likewise, we will
let $V_1$ and $V_2$ denote two distinct alleles at a variable marker.  We can then describe
an individual's genotype as $V_1V_1$ or $V_1V_2$.  In the above, where we talk about
"variable alleles assigned to $G,G'$" we are talking about the possible distinct assignments
of variable alleles to the gamete that gets segregated and the one that remains.
Obviously, if an $H_1H_2$ individual that is also $V_1V_2$ segregates an $H_1$ then
there are two possible assignments of the variable alleles: either the $V_1$ gets
segregated to the offspring upon the $H_1$ background and the $V_2$ is on the
$H_2$ background, or the $V_2$ gets segregated on the $H_1$ background and $V_1$ is
on the $H_2$ background.  


The $\mbox{Prob(parent genotype given }v\mbox{ on }G\mbox{ and }G')$ is fairly straightforward.
Once we assign alleles to ancestries (which we do when we assign the allele to
the segregated ancestry), then the probability of the parent's genotype is merely
the chance of drawing each of the alleles in the individual given the ancestries/subspecies
they are from.  These are just the allele frequencies that we calculated using the
EM-algorithm, above.

Likewise, the $\mbox{Prob(offsping genotype given }v\mbox{ on }G\mbox{ was segregated to it})$ is
pretty easily calculated.  Let's denote the segregated gamete as $H_1,V_1$ ($H_1$ is simply
the ancestry and $V_1$ is the allele).  Then, keeping in mind that we are only dealing with
_compatible_ intersected intervals, the possibilities for the offspring genotypes
and their corresponding probabilities are:
$$
\begin{aligned}
H_1H_1,V_1V_1 & & p_{H_1}(V_1)\\
H_1H_1,V_1V_2 & & p_{H_1}(V_2) \\
H_1H_1,V_2V_2 & & \epsilon p_{H_1}(V_2) \\
H_1H_1,V_2V_3 & & \epsilon p_{H_1}(V_2)/2 + \epsilon p_{H_1}(V_3)/2 \\
H_1H_2,V_1V_1 & & p_{H_2}(V_1) \\
H_1H_2,V_2V_2 & & \epsilon p_{H_2}(V_2)  \\
H_1H_2,V_2V_3 & & \epsilon p_{H_2}(V_2)/2 + \epsilon p_{H_2}(V_3)/2  \\
\end{aligned}
$$
where $\epsilon$ is essentially the rate at which alleles at the variable loci
are miscalled, in a very
simple genotyping error model.  

Great! All I need now is a little pseudo code for the products and sums above
and I will be ready to start implementing this.  I will do it using RCpp because,
when we are doing this sort of thing between many pairs of individuals, that
ends up being a lot of computing, and RCpp will be considerably faster.

Actually, I am just going to write up a big table for these.  Here is how we do it.  Since
we are dealing just with SNPs that are biallelic, and we are going to filter down to just
two species/ancestries, so we have $A_2B_2C_0$, $A_1B_1C_0$, and $A_0B_2C_0$.  And the
genotypes will be denoted at 0, 1, or 2 (the number of occurrences of the "1" allele). Note
that the allele frequencies, $p_A$ and $p_B$ are given as the frequency of the 1 allele on the
$A$ or the $B$ backgrounds, respectively. Genotyping errors are dealt with via the super simple
model: we assume that the parent genotype is always correctly identified, while the
offspring genotype is incorrect with probability $\epsilon$, in which case the
genotype of the offspring just gets drawn from the genotype probs given its ancestry.
(This has the effect that the ratio becomes $\epsilon$).
\begin{table}
\caption{Relevant genotype probabilities when the parent is $A_2B_0C_0$.
$H$ is the ancestry pattern specified according to the values of the trits when
the integer is expressed in ternary form ($A_2B_0C_0 = 2$; $A_1B_1C_0 = 4$).
$\opro_X$ is the probability of receiving a segment from species $X$ from the "other"
parent---i.e., the one that is not the candidate. The $\opro$'s are, by and large, hard
to model, because it depends on the distribution of admixture fractions/patterns amongst
the possible parents in the population, and they are not independent across markers (because
they all came from a single parent).  I have considered urn-model-like ways of modeling them,
but I think I might just end up setting them all to 1, so that they don't affect the probability.
At any rate, the main thing here is that the ultimate probabilities we want for our likelihood
ratios are not simply the $P(G_\kid|G_\pa, H_\pa, H_\kid, p_A, p_B)$ probabilities, but actually
we are going to want,
the $P(H_\kid,G_\kid|G_\pa, H_\pa, p_A, p_B) = P(G_\kid|G_\pa, H_\pa, H_\kid, p_A, p_B)P(H_\kid|H_\pa)$,
in the numerator and 
$P(H_\kid,G_\kid| H_\kid,p_A, p_B) = P(G_\kid| H_\kid,p_A, p_B)P(H_\kid)$.  Note that $P(H_\kid)$ would be,
for example, $\opro_A^2$ if $H_\kid=2$, under the simplest independence assumptions.  So, if we just
set all the $\opro$'s to 1, that would work out fine.  Hot Damn!  This is starting to make sense (maybe...).
}
$$
\small
\begin{aligned}
H_\pa   & & H_\kid & & G_\pa   & & G_\kid  & &P(H_\kid|H_\pa)&  & P(G_\kid|G_\pa, H_\pa, H_\kid, p_A, p_B) &~~~& P(G_\kid| H_\kid,p_A, p_B)\\ \hline
\\
2       & & 2      & &  0      & &  0      & & \opro_A          &   &  (1-\epsilon)(1-p_A) + \epsilon(1-p_A)^2                   &   &  (1-p_A)^2 \\
2       & & 2      & &  0      & &  1      & & \opro_A           &  &  (1-\epsilon)p_A + \epsilon 2p_A(1-p_A)                      &   &  2p_A(1-p_A) \\
2       & & 2      & &  0      & &  2      & & \opro_A &  &  \epsilon p_A^2             &   &  p_A^2 \\
\\
2       & & 2      & &  1      & &  0      & & \opro_A &  &  (1-\epsilon)\frac{1}{2}(1-p_A) + \epsilon(1-p_A)^2                   &   &  (1-p_A)^2 \\
2       & & 2      & &  1      & &  1      & & \opro_A &  &  (1-\epsilon)\frac{1}{2} + \epsilon 2p_A(1-p_A)                       &   &  2p_A(1-p_A) \\
2       & & 2      & &  1      & &  2      & & \opro_A &  &  (1-\epsilon)\frac{1}{2}p_A + \epsilon p_A^2             &   &  p_A^2 \\
\\
2       & & 2      & &  2      & &  0      & & \opro_A &  &  \epsilon(1-p_A)^2                    &   &  (1-p_A)^2 \\
2       & & 2      & &  2      & &  1      & & \opro_A &  &  (1-\epsilon)(1-p_A) + \epsilon 2p_A(1-p_A)             &   &  2p_A(1-p_A) \\
2       & & 2      & &  2      & &  2      & & \opro_A &  &  (1-\epsilon) p_A  + \epsilon p_A^2            &   &  p_A^2 \\
\\
2       & & 4      & &  0      & &  0      & & \opro_B &  &  (1-\epsilon)(1-p_B) + \epsilon(1-p_A)(1-p_B)         &   &  (1-p_A)(1-p_B) \\
2       & & 4      & &  0      & &  1      & & \opro_B &  &  (1-\epsilon)p_B + \epsilon[p_A(1-p_B) + p_B(1-p_A)]         &   &  p_A(1-p_B) + p_B(1-p_A) \\
2       & & 4      & &  0      & &  2      & & \opro_B &  &  \epsilon p_Ap_B        &   &  p_Ap_B \\
\\
2       & & 4      & &  1      & &  0      & & \opro_B &  &  (1-\epsilon)\frac{1}{2}(1-p_B) + \epsilon(1-p_A)(1-p_B)        &   &  (1-p_A)(1-p_B) \\
2       & & 4      & &  1      & &  1      & & \opro_B &  &  (1-\epsilon)\frac{1}{2} + \epsilon[p_A(1-p_B) + p_B(1-p_A)]         &   &  p_A(1-p_B) + p_B(1-p_A) \\
2       & & 4      & &  1      & &  2      & & \opro_B &  &  (1-\epsilon)\frac{1}{2}p_B + \epsilon p_Ap_B        &   &  p_Ap_B \\
\\
2       & & 4      & &  2      & &  0      & & \opro_B &  &  \epsilon(1-p_A)(1-p_B)                          &   &  (1-p_A)(1-p_B) \\
2       & & 4      & &  2      & &  1      & & \opro_B &  &  (1-\epsilon)(1-p_B) + \epsilon[p_A(1-p_B) + p_B(1-p_A)]         &   &  p_A(1-p_B) + p_B(1-p_A) \\
2       & & 4      & &  2      & &  2      & & \opro_B &  &  (1-\epsilon)p_B + \epsilon p_Ap_B        &   &  p_Ap_B \\
\end{aligned}
$$
\end{table}


\begin{table}
\caption{SAME TABLE BUT EXPRESSED DIFFERENTLY. Here the $P^*$ term assumes no genotyping
error.  And we know that we multiple that by $(1-\epsilon)$ and add $\epsilon$ times the
kid's marginal genotype probability to get what we need. Dependence on $p_A$ and $p_B$
is implicit.  Relevant genotype probabilities when the parent is $A_2B_0C_0$.
$H$ is the ancestry pattern specified according to the values of the trits when
the integer is expressed in ternary form ($A_2B_0C_0 = 2$; $A_1B_1C_0 = 4$).
$\opro_X$ is the probability of receiving a segment from species $X$ from the "other"
parent---i.e., the one that is not the candidate. The $\opro$'s are, by and large, hard
to model, because it depends on the distribution of admixture fractions/patterns amongst
the possible parents in the population, and they are not independent across markers (because
they all came from a single parent).  I have considered urn-model-like ways of modeling them,
but I think I might just end up setting them all to 1, so that they don't affect the probability.
At any rate, the main thing here is that the ultimate probabilities we want for our likelihood
ratios are not simply the $P(G_\kid|G_\pa, H_\pa, H_\kid, p_A, p_B)$ probabilities, but actually
we are going to want,
the $P(H_\kid,G_\kid|G_\pa, H_\pa, p_A, p_B) = P(G_\kid|G_\pa, H_\pa, H_\kid, p_A, p_B)P(H_\kid|H_\pa)$,
in the numerator and 
$P(H_\kid,G_\kid| H_\kid,p_A, p_B) = P(G_\kid| H_\kid,p_A, p_B)P(H_\kid)$.  Note that $P(H_\kid)$ would be,
for example, $\opro_A^2$ if $H_\kid=2$, under the simplest independence assumptions.  So, if we just
set all the $\opro$'s to 1, that would work out fine.  Hot Damn!  This is starting to make sense (maybe...).
}
$$
\small
\begin{aligned}
H_\pa   & & H_\kid & & G_\pa   & & G_\kid  & &P(H_\kid|H_\pa)&  & P^*(G_\kid|G_\pa, H_\pa, H_\kid) &~~~& P(G_\kid| H_\kid)\\ \hline
\\
2       & & 2      & &  0      & &  0      & & \opro_A          &   &  1-p_A                   &   &  (1-p_A)^2 \\
2       & & 2      & &  0      & &  1      & & \opro_A           &  &  p_A                     &   &  2p_A(1-p_A) \\
2       & & 2      & &  0      & &  2      & & \opro_A &  &                     0             &   &  p_A^2 \\
\\
2       & & 2      & &  1      & &  0      & & \opro_A &  &  \frac{1}{2}(1-p_A)                   &   &  (1-p_A)^2 \\
2       & & 2      & &  1      & &  1      & & \opro_A &  &  \frac{1}{2}                          &   &  2p_A(1-p_A) \\
2       & & 2      & &  1      & &  2      & & \opro_A &  &  \frac{1}{2}p_A                  &   &  p_A^2 \\
\\
2       & & 2      & &  2      & &  0      & & \opro_A &  &  0                    &   &  (1-p_A)^2 \\
2       & & 2      & &  2      & &  1      & & \opro_A &  &  1-p_A             &   &  2p_A(1-p_A) \\
2       & & 2      & &  2      & &  2      & & \opro_A &  &  p_A            &   &  p_A^2 \\
\\
2       & & 4      & &  0      & &  0      & & \opro_B &  &  1-p_B         &   &  (1-p_A)(1-p_B) \\
2       & & 4      & &  0      & &  1      & & \opro_B &  &  p_B           &   &  p_A(1-p_B) + p_B(1-p_A) \\
2       & & 4      & &  0      & &  2      & & \opro_B &  &  0             &   &  p_Ap_B \\
\\
2       & & 4      & &  1      & &  0      & & \opro_B &  &  \frac{1}{2}(1-p_B)        &   &  (1-p_A)(1-p_B) \\
2       & & 4      & &  1      & &  1      & & \opro_B &  &  \frac{1}{2}         &   &  p_A(1-p_B) + p_B(1-p_A) \\
2       & & 4      & &  1      & &  2      & & \opro_B &  &  \frac{1}{2}p_B      &   &  p_Ap_B \\
\\
2       & & 4      & &  2      & &  0      & & \opro_B &  &  0                          &   &  (1-p_A)(1-p_B) \\
2       & & 4      & &  2      & &  1      & & \opro_B &  &  1-p_B         &   &  p_A(1-p_B) + p_B(1-p_A) \\
2       & & 4      & &  2      & &  2      & & \opro_B &  &  p_B           &   &  p_Ap_B \\
\end{aligned}
$$
\end{table}




\begin{table}
\caption{Relevant genotype probabilities when the parent is $A_1B_1C_0$. Note that when there
are only two species, then $\opro_A + \opro_B = 1$.  Of course, that sort of makes things weird
if we set each of them to 1. Maybe we set them each to 1/2?
And for compactness, I think we can define, $f_{AB}(01) = \frac{(1-p_A)p_B}{(1-p_A)p_B + p_A(1-p_B)}$
to be the probability that the 0 allele is on the $A$ background (and the 1 allele is on the $B$
background) in the parent, when it is heterozygous. Thus,
$f_{AB}(10) = \frac{p_A(1-p_B)}{(1-p_A)p_B + p_A(1-p_B)}$ is the probability that the
1 allele is on $A$ and the 0 allele is on $B$ within that parent. Hey! I should
make dependence on $p_A$ and $p_B$ implicit to save some space in the headers.
}
$$
\footnotesize
\begin{aligned}
H_\pa   & & H_\kid & & G_\pa   & & G_\kid  & &P(H_\kid|H_\pa)&  & P^*(G_\kid|G_\pa, H_\pa, H_\kid) &~~~& P(G_\kid| H_\kid)\\ \hline
\\
4       & & 2      & &  0      & &  0      & & \frac{1}{2}\opro_A         &   &  1-p_A                   &   &  (1-p_A)^2 \\
4       & & 2      & &  0      & &  1      & & \frac{1}{2}\opro_A           &  &  p_A                       &   &  2p_A(1-p_A) \\
4       & & 2      & &  0      & &  2      & & \frac{1}{2}\opro_A &  &       0             &   &  p_A^2 \\
\\
4       & & 2      & &  1      & &  0      & & \frac{1}{2}\opro_A  &  &  f_{AB}(01)(1-p_A)                   &   &  (1-p_A)^2 \\
4       & & 2      & &  1      & &  1      & & \frac{1}{2}\opro_A  &  &  f_{AB}(01)p_A + f_{AB}(10)(1-p_A)               &   &  2p_A(1-p_A) \\
4       & & 2      & &  1      & &  2      & & \frac{1}{2}\opro_A  &  &  f_{AB}(10)p_A                     &   &  p_A^2 \\
\\
4       & & 2      & &  2      & &  0      & & \frac{1}{2}\opro_A  &  &  0                    &   &  (1-p_A)^2 \\
4       & & 2      & &  2      & &  1      & & \frac{1}{2}\opro_A  &  &  1-p_A             &   &  2p_A(1-p_A) \\
4       & & 2      & &  2      & &  2      & & \frac{1}{2}\opro_A  &  &  p_A               &   &  p_A^2 \\
\\
4       & & 4      & &  0      & &  0      & & \frac{1}{2}(\opro_A + \opro_B) &  &  \frac{1}{2}(1-p_B) + \frac{1}{2}(1-p_A)         &   &  (1-p_A)(1-p_B) \\
4       & & 4      & &  0      & &  1      & & \frac{1}{2}(\opro_A + \opro_B) &  &  \frac{1}{2}p_B + \frac{1}{2}p_A          &   &  p_A(1-p_B) + p_B(1-p_A) \\
4       & & 4      & &  0      & &  2      & & \frac{1}{2}(\opro_A + \opro_B) &  &        0          &   &  p_Ap_B \\
\\
4       & & 4      & &  1      & &  0      & & \frac{1}{2}(\opro_A + \opro_B) &  &  \frac{1}{2}f_{AB}(01)(1-p_B) +  \frac{1}{2}f_{AB}(10)(1-p_A)      &   &  (1-p_A)(1-p_B) \\
4       & & 4      & &  1      & &  1      & & \frac{1}{2}(\opro_A + \opro_B) &  &  \frac{1}{2}[f_{AB}(01)(1-pA+p_B) +f_{AB}(10)(1-p_B + p_A)]   &   &  p_A(1-p_B) + p_B(1-p_A) \\
4       & & 4      & &  1      & &  2      & & \frac{1}{2}(\opro_A + \opro_B) &  &  \frac{1}{2}f_{AB}(10)p_B +  \frac{1}{2}f_{AB}(01)p_A      &   &  p_Ap_B \\
\\
4       & & 4      & &  2      & &  0      & & \frac{1}{2}(\opro_A + \opro_B) &  &  0                          &   &  (1-p_A)(1-p_B) \\
4       & & 4      & &  2      & &  1      & & \frac{1}{2}(\opro_A + \opro_B) &  &  \frac{1}{2}(1-p_A) + \frac{1}{2}(1-p_B)        &   &  p_A(1-p_B) + p_B(1-p_A) \\
4       & & 4      & &  2      & &  2      & & \frac{1}{2}(\opro_A + \opro_B) &  &   \frac{1}{2}p_B + \frac{1}{2}p_A       &   &  p_Ap_B \\
\end{aligned}
$$
\end{table}
